src/backend/distributed/commands/create_distributed_table.c

```diff
@@ -810,53 +899,66 @@ CopyLocalDataIntoShards(Oid distributedRelationId)
 	econtext->ecxt_scantuple = slot;
 
 	copyDest =
-		(DestReceiver *) CreateCitusCopyDestReceiver(distributedRelationId,
+		(DestReceiver *) CreateCitusCopyDestReceiver(destinationDistributedRelationId,
 													 columnNameList, estate,
 													 stopOnFailure);
 
 	/* initialise state for writing to shards, we'll open connections on demand */
 	copyDest->rStartup(copyDest, 0, tupleDescriptor);
 
-	/* begin reading from local table */
-	scan = heap_beginscan(distributedRelation, GetActiveSnapshot(), 0, NULL);
-
-	oldContext = MemoryContextSwitchTo(GetPerTupleMemoryContext(estate));
-
-	while ((tuple = heap_getnext(scan, ForwardScanDirection)) != NULL)
+	/* iterate on the local relations, copy each into the destination relation */
+	foreach(sourceRelationCell, sourceRelationList)
```

Comment:
> However, you should be able to call CreateHashDistributedTable for all tables, and then CopyLocalDataIntoShards for all tables.

Yes, I did exactly the above on this [branch](https://github.com/citusdata/citus/compare/partition_distributed_table_create_table...connection_error?expand=1) but the behavior is still the same. I'll debug more and share my understanding on what's happening.